README.md
![Build](https://github.com/koovee/thermia/actions/workflows/build.yml/badge.svg)
![Test](https://github.com/koovee/thermia/actions/workflows/test.yml/badge.svg)

# Thermia controller

This project controls Thermia Diplomat Duo heat pump using Shelly 1 plus relay.

Thermia:

- short 307 and 308 pins: *EVU STOP*
- use 10 kOhm resistance between 307 and 308: *ROOM LOWERING* mode

Shelly switch is connected between 307 and 308 pins.

There are two operating modes: *simple threshold* and *dynamic threshold*.

## Threshold

Heating is controlled based on `THRESHOLD` (maximum price in *c/kWh*). Heating is *OFF* or in *ROOM LOWERING* mode 
when price is higher than threshold and ON when price is lower than threshold.

## Active hours

Heating is controlled based on `ACTIVE_HOURS` (specifies the number of hours that heating must be on during a day).
Heating is *OFF* or in *ROOM LOWERING* mode when current hour is not one of the `ACTIVE_HOURS` cheapest hours of the 
day.

## Threshold and active hours

Heating is on if hour price is lower than the `THRESHOLD` or hour is one of the cheapest hours of the day. This
makes sure that heating is on at least *n* hours a day. Number of hours is specified by `ACTIVE_HOURS` environment 
variable.

## Schedule

This is fallback mode that is normally used when *spot price* information is not available. Default hours are 00-06. 
This can be overriden with `SCHEDULE` environment variable.

```
# from 00 to 06
SCHEDULE="00,01,02,03,04,05,06"
```

In case only `SCHEDULE` environment variable is specified, fallback mode is used.

# Configuration

`THRESHOLD` maximum price (*c/kWh*). Heating is ON if *spot price* is lower and OFF if *spot price* is higher than the
threshold.

`ACTIVE_HOURS` number of hours that heating must be ON





CHANGELOG.md
# CHANGELOG

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/) and [Keep a Changelog](http://keepachangelog.com/).



## Unreleased
---

### New

### Changes

### Fixes

### Breaks

## 0.3.0 - (2022-10-28)
---

### New
* add maxPrice setting (heating off/lower if price is higher than maxPrice)

### Fixes
* day ahead price handling

## 0.2.0 - (2022-10-28)
---

### New
* add fallback mode (schedule)


## 0.1.0 - (2022-10-26)
---

### New
* add docker build, fix tests, add timezone support
* Threshold can be overridden by setting activeHours
* add -dryrun cmdline option (does not control relay when this is set)

### Changes
* use interfaces

### Fixes
* use 0.00 price when pricing is not available




src/control/control_test.go
package control

import (
	"os"
	"testing"
)

func TestMain(m *testing.M) {
	os.Exit(m.Run())
}

func TestInit(t *testing.T) {
	s := State{}

	if err := s.Init(false); err != nil {
		t.Errorf("init() with defaults did not succeed")
	}

	if err := s.Init(true); err != nil {
		t.Errorf("init() with defaults did not succeed")
	}

	os.Setenv("SHELLY_URL", "http://127.0.0.1")
	if err := s.Init(false); err != nil {
		t.Errorf("init() did not succeed")
	}
	os.Unsetenv("SHELLY_URL")
}

// Need to mock http client..
//func TestSwitchOn(t *testing.T) {
//	cases := map[string]struct {
//		expectedResult error
//	}{
//		"SwitchON when Switch is OFF": {
//			expectedResult: nil,
//		},
//	}
//
//	os.Setenv("SHELLY_URL", "http://")
//	s := State{}
//	s.Init()
//
//	for k, tc := range cases {
//		err := s.SwitchOn()
//		if err != tc.expectedResult {
//			t.Fatalf("%s: SwtichOn\ngot:  %s\nwant: %s\n", k, err, tc.expectedResult)
//		}
//	}
//}


src/spotprice/spotprice_test.go
package spotprice

import (
	"os"
	"testing"
	"time"
)

func TestMain(m *testing.M) {
	os.Exit(m.Run())
}

func TestInit(t *testing.T) {
	s := State{}

	if err := s.Init(); err == nil {
		t.Errorf("init() should have failed, but it succeeded")
	}

	os.Setenv("TOKEN", "12345")
	if err := s.Init(); err != nil {
		t.Errorf("init() with TOKEN set did not succeed")
	}
	os.Unsetenv("TOKEN")
}

func TestCheapestHours(t *testing.T) {
	s := State{}
	s.HourPrice = make(map[string][]float64)
	set1 := []float64{1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0, 23.0, 24.0}
	set2 := []float64{-5.0, -4.0, -3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0}

	cases := map[string]struct {
		hourPrice      []float64
		hour           int
		hours          int
		expectedResult bool
	}{
		"Set1: Is hour 0 the cheapest hour":            {hourPrice: set1, hour: 0, hours: 1, expectedResult: true},
		"Set1: Is hour 1 the cheapest hour":            {hourPrice: set1, hour: 1, hours: 1, expectedResult: false},
		"Set1: Is hour 4 on of the 5 cheapest hours":   {hourPrice: set1, hour: 4, hours: 5, expectedResult: true},
		"Set1: Is hour 4 on of the 4 cheapest hours":   {hourPrice: set1, hour: 4, hours: 4, expectedResult: false},
		"Set1: Is hour 23 on of the 24 cheapest hours": {hourPrice: set1, hour: 23, hours: 24, expectedResult: true},
		"Set1: Is hour 23 on of the 23 cheapest hours": {hourPrice: set1, hour: 23, hours: 23, expectedResult: false},

		"Set2: Is hour 0 the cheapest hour":           {hourPrice: set2, hour: 0, hours: 1, expectedResult: true},
		"Set2: Is hour 1 the cheapest hour":           {hourPrice: set2, hour: 1, hours: 1, expectedResult: false},
		"Set2: Is hour 10 on of the 5 cheapest hours": {hourPrice: set2, hour: 10, hours: 5, expectedResult: false},
	}

	for k, tc := range cases {
		s.HourPrice[time.Now().Format(DateLayout)] = tc.hourPrice
		result := IsCheapestHour(tc.hour, s.CheapestHours(tc.hours))
		if result != tc.expectedResult {
			t.Fatalf("%s: IsCheapestHour\ngot:  %v\nwant: %v\n", k, result, tc.expectedResult)
		}
	}
}


src/control/shelly.go
package control

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
)

const (
	defaultShellyUrl = "http://10.0.0.84/relay/0"
)

type State struct {
	url    string
	hc     *http.Client
	dryRun bool
}

type statusResponse struct {
	Ison           bool    `json:"ison"`
	HasTimer       bool    `json:"has_timer"`
	TimerStartedAt int     `json:"timer_started_at"`
	TimerDuration  float64 `json:"timer_duration"`
	TimerRemaining float64 `json:"timer_remaining"`
	Source         string  `json:"source"`
}

var s State

func (s *State) Init(dryRun bool) error {
	err := s.getEnv()
	if err != nil {
		fmt.Printf("failed to get required environment variables")
		return err
	}
	s.dryRun = dryRun
	return nil
}

// SwitchOff turns switch OFF which means Thermia is operating in NORMAL mode
func (s State) SwitchOff() error {
	var response statusResponse

	// check current state
	resp, err := http.Get(s.url)
	if err != nil {
		fmt.Printf("Failed to create http request")
		return errors.New("failed to create http request")
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)                  // response body is []byte
	if err := json.Unmarshal(body, &response); err != nil { // Parse []byte to go struct pointer
		fmt.Printf("Can not unmarshal JSON: %s\n", err.Error())
		return errors.New("failed to unmarshal JSON response")
	}
	if response.Ison == true {
		// change state
		if s.dryRun {
			fmt.Printf("DRY RUN -- Switch is on, turning it off (NORMAL OPERATION) -- DRY RUN\n")
		} else {
			fmt.Printf("Switch is on, turning it off (NORMAL OPERATION)\n")
			resp, err := http.Get(s.url + "?turn=off")
			if err != nil {
				fmt.Printf("Failed to create http request")
				return errors.New("failed to create http request")
			}
			defer resp.Body.Close()
			if resp.StatusCode != http.StatusOK {
				fmt.Printf("failed to set switch off: %s\n", err.Error())
				return errors.New("failed to set switch off")
			}
		}
	}

	return nil
}

// SwitchOn tunrs switch ON which means Thermia is operating in heat reduction mode (normal-2 degress)
func (s State) SwitchOn() error {
	var response statusResponse

	// check current state
	resp, err := http.Get(s.url)
	if err != nil {
		fmt.Printf("Failed to create http request")
		return errors.New("failed to create http request")
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)                  // response body is []byte
	if err := json.Unmarshal(body, &response); err != nil { // Parse []byte to go struct pointer
		fmt.Printf("Can not unmarshal JSON: %s\n", err.Error())
		return errors.New("failed to unmarshal JSON response")
	}
	if response.Ison == false {
		// change state
		if s.dryRun {
			fmt.Printf("DRY RUN -- Switch is off, turning it on (EVU ON / LOWERED TEMPERATURE) -- DRY RUN\n")
		} else {
			fmt.Printf("Switch is off, turning it on (EVU ON / LOWERED TEMPERATURE)\n")
			resp, err := http.Get(s.url + "?turn=on")
			if err != nil {
				fmt.Printf("Failed to create http request")
				return errors.New("failed to create http request")
			}
			defer resp.Body.Close()
			if resp.StatusCode != http.StatusOK {
				fmt.Printf("failed to set switch on: %s\n", err.Error())
				return errors.New("failed to set switch on")
			}
		}
	}

	return nil
}

func (s *State) getEnv() error {
	s.url = os.Getenv("SHELLY_URL")
	if s.url == "" {
		s.url = defaultShellyUrl
	}
	return nil
}


src/spotprice/entsoe.go
package spotprice

import (
	"encoding/xml"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"sync"
	"time"
)

const (
	apiUrl     = "https://transparency.entsoe.eu/api"
	DateLayout = "20060102"
	highPrice  = 9999.99
)

type State struct {
	token     string
	threshold float64
	maxPrice  float64
	HourPrice HourPrices
	C         chan bool
	M         *sync.Mutex
	hc        http.Client
}

type A44Response struct {
	XMLName    xml.Name `xml:"Publication_MarketDocument"`
	TimeSeries []struct {
		XMLName xml.Name `xml:"TimeSeries"`
		MRID    string   `xml:"mRID"`
		Period  struct {
			XMLName      xml.Name `xml:"Period"`
			TimeInterval struct {
				XMLName xml.Name `xml:"timeInterval"`
				Start   string   `xml:"start"`
				End     string   `xml:"end"`
			}
			Resolution string `xml:"resolution"`
			Point      []struct {
				XMLName  xml.Name `xml:"Point"`
				Position int      `xml:"position"`
				Price    string   `xml:"price.amount"`
			}
		}
	}
}

func (s *State) Init() (err error) {
	err = s.getEnv()
	if err != nil {
		fmt.Printf("failed to get required environment variables: %s\n", err.Error())
		return err
	}

	s.hc = http.Client{
		Transport:     nil,
		CheckRedirect: nil,
		Jar:           nil,
		Timeout:       0,
	}

	s.C = make(chan bool)
	s.M = &sync.Mutex{}
	s.HourPrice = make(map[string][]float64)

	return nil
}

// GetPrice returns price in c/kWh
func (s State) GetPrice(time time.Time) (float64, error) {
	s.M.Lock()
	defer s.M.Unlock()
	hour := time.Hour()
	if len(s.HourPrice[time.Format(DateLayout)]) < hour {
		fmt.Printf("no pricing available for %s hour %d\n", time.String(), hour)
		return 0, errors.New("no price information available")
	}
	return s.HourPrice[time.Format(DateLayout)][hour] / 10, nil
}

// UpdateSpotPrices ..
func (s *State) UpdateSpotPrices() {
	var retryCount = 0

	yesterday := time.Now().Add(-24 * time.Hour).Format(DateLayout)
	day := time.Now().Format(DateLayout)
	tomorrow := time.Now().Add(24 * time.Hour).Format(DateLayout)

	s.M.Lock()
	defer s.M.Unlock()

	periodStart := day + "0000"
	periodEnd := tomorrow + "0000"

	if time.Now().Hour() > 18 && len(s.HourPrice[day]) > 0 {
		if len(s.HourPrice[tomorrow]) < 24 {
			periodStart = tomorrow + "0000"
			periodEnd = tomorrow + "0100"
			day = tomorrow
			// TODO: change variable name...
		} else {
			// enough pricing data in store..
			return
		}
	} else if len(s.HourPrice[day]) > 0 {
		// enough pricing data in store..
		return
	}

	fmt.Printf("getting spot prices from %s\n", apiUrl)

	// delete yesterdays records
	fmt.Printf("DEBUG: map size before cleanup: %d\n", len(s.HourPrice))
	delete(s.HourPrice, yesterday)
	fmt.Printf("DEBUG: map size after cleanup: %d\n", len(s.HourPrice))

	req, err := http.NewRequest("GET", apiUrl, nil)
	if err != nil {
		fmt.Printf("Failed to create http request\n")
		retryCount++
		time.Sleep(time.Second * time.Duration(retryCount*retryCount))
		return
	}
	retryCount = 0

	q := url.Values{}
	q.Add("securityToken", s.token)
	q.Add("documentType", "A44")
	q.Add("In_domain", "10YFI-1--------U")
	q.Add("out_domain", "10YFI-1--------U")
	q.Add("periodStart", periodStart)
	q.Add("periodEnd", periodEnd)
	req.URL.RawQuery = q.Encode()

	resp, err := s.hc.Do(req)
	if err != nil {
		fmt.Printf("failed to make http request: %s\n", err.Error())
		return
	}
	defer resp.Body.Close()

	fmt.Printf("DEBUG: %s\n", req.URL)
	fmt.Printf("DEBUG: %v\n", resp.Status)

	hourlyPrices := A44Response{}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Printf("failed to read http response body\n")
		return
	}
	//fmt.Printf("body: %s\n", body)

	if err = xml.Unmarshal(body, &hourlyPrices); err != nil {
		fmt.Printf("failed to unmarshal xml\n")
		return
	}

	for _, v := range hourlyPrices.TimeSeries {
		var p []float64

		periodStart, err := time.Parse("2006-01-02T15:04Z", v.Period.TimeInterval.Start)
		if err != nil {
			fmt.Printf("failed to parse Period.TimeInterval.Start: %s\n", err.Error())
			fmt.Printf("DEBUG response body: %s\n", body)
			return
		}
		day = periodStart.Add(time.Hour * 24).Format(DateLayout)
		tomorrow = periodStart.Add(time.Hour * 48).Format(DateLayout)

		// add dummy data for offset from utc hours
		_, offset := time.Now().Zone()
		offsetHours := offset / 3600
		for i := 0; i < offset/3600; i++ {
			p = append(p, highPrice)
		}

		for _, v := range v.Period.Point {
			if v.Position > 24 {
				fmt.Printf("DEBUG: v.Position larger than 24 -- check entsoe response!!\n")
				fmt.Printf("DEBUG response body: %s\n", body)
			} else {
				price, err := strconv.ParseFloat(v.Price, 64)
				if err != nil {
					fmt.Printf("failed to convert price to float\n")
					return
				}
				p = append(p, price)
			}
		}
		if len(s.HourPrice[day]) == 0 {
			s.HourPrice[day] = p[offsetHours-1 : 24+offsetHours-1]
			s.HourPrice[tomorrow] = p[24+offsetHours-1:]
		} else {
			// length is 1
			s.HourPrice[day] = append(s.HourPrice[day], p[offsetHours:24+offsetHours-1]...)
			s.HourPrice[tomorrow] = p[24+offsetHours-1:]
		}
	}

	fmt.Printf("DEBUG: %v\n", s.HourPrice)
}

// CheapestHours returns the cheapest n hours for a given day
func (s State) CheapestHours(n int) (cheapestPrices []int) {
	var cheapestIndex int
	var cheapest float64

	prices := s.HourPrice[time.Now().Format(DateLayout)]

	for i := 0; i < n; i++ {
		cheapest = highPrice
		for i, price := range prices {
			if price < cheapest {
				if len(cheapestPrices) > 0 {
					if price <= prices[cheapestPrices[len(cheapestPrices)-1]] {
						continue
					}
				}
				cheapest = price
				cheapestIndex = i
			}
		}
		cheapestPrices = append(cheapestPrices, cheapestIndex)
	}
	return cheapestPrices
}

func (s *State) getEnv() error {
	s.token = os.Getenv("TOKEN")
	if s.token == "" {
		return errors.New("TOKEN not set")
	}
	return nil
}


src/main.go
package main

import (
	"errors"
	"flag"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/koovee/thermia/control"
	"github.com/koovee/thermia/spotprice"
)

const (
	defaultTimezone = "Europe/Helsinki"
	defaultSchedule = "0,1,2,3,4,5"
)

var version string

type state struct {
	sp          spotprice.State
	cs          control.State
	threshold   float64
	maxPrice    float64
	activeHours int
	schedule    map[int]bool
	tz          string
}

func main() {

	dryRun := flag.Bool("dryrun", false, "disable relay control")
	flag.Parse()

	s, err := getEnv()
	if err != nil {
		fmt.Printf("failed to get required environment variables\n")
		return
	}

	_, err = time.LoadLocation(s.tz)
	if err != nil {
		fmt.Printf("failed to set timezone (%s): %s\n", s.tz, err.Error())
		return
	}

	err = s.sp.Init()
	if err != nil {
		fmt.Printf("failed to initialize spotprice module\n")
		return
	}
	err = s.cs.Init(*dryRun)
	if err != nil {
		fmt.Printf("failed to initialize shelly module\n")
		return
	}

	fmt.Printf("Thermia controller started (version: %s, dryRun: %v, treshold: %0.2f, activeHours: %d)\n", version, *dryRun, s.threshold, s.activeHours)

	// every even hour
	timer := time.NewTimer(time.Second)

	for {
		select {
		case <-timer.C:
			// Update prices
			s.sp.UpdateSpotPrices()

			// Control relay based on configuration and hourly price
			if s.activeHours > 0 && s.threshold > 0 {
				err = s.controlBasedOnThresholdAndActiveHours()
				if err != nil {
					err = s.controlBasedOnSchedule()
				}
			} else if s.activeHours > 0 {
				err = s.controlBasedOnActiveHours()
				if err != nil {
					err = s.controlBasedOnSchedule()
				}
			} else if s.threshold > 0 {
				err = s.controlBasedOnThreshold()
				if err != nil {
					err = s.controlBasedOnSchedule()
				}
			} else {
				err = s.controlBasedOnSchedule()
			}
			if err != nil {
				fmt.Printf("failed to control relay: %s\n", err.Error())
			}

			timer.Reset(time.Now().Truncate(time.Hour).Add(time.Hour).Add(time.Second).Sub(time.Now()))
		}
	}
}

func getEnv() (s state, err error) {
	threshold := os.Getenv("THRESHOLD")
	if threshold != "" {
		s.threshold, err = strconv.ParseFloat(threshold, 64)
		if err != nil {
			fmt.Printf("failed to parse float from environment variable (THRESHOLD): %s\n", err.Error())
			return
		}
	}

	maxPrice := os.Getenv("MAX_PRICE")
	if maxPrice != "" {
		s.maxPrice, err = strconv.ParseFloat(maxPrice, 64)
		if err != nil {
			fmt.Printf("failed to parse float from environment variable (MAX_PRICE): %s\n", err.Error())
			return
		}
	}

	activeHours := os.Getenv("ACTIVE_HOURS")
	if activeHours != "" {
		s.activeHours, err = strconv.Atoi(activeHours)
		if err != nil {
			fmt.Printf("failed to parse int from environment variable (ACTIVE_HOURS): %s\n", err.Error())
			return
		}
	}

	schedule := os.Getenv("SCHEDULE")
	if schedule == "" {
		schedule = defaultSchedule
	}

	s.schedule = make(map[int]bool)
	scheduleString := strings.Split(schedule, ",")
	for _, str := range scheduleString {
		var hour int
		hour, err = strconv.Atoi(str)
		if err != nil {
			fmt.Printf("failed to parse array of integers from environment variable (SCHEDULE): %s\n", err.Error())
			return
		}
		if hour < 0 || hour > 24 {
			err = errors.New("invalid hour")
			fmt.Printf("failed to parse array of integers from environment variable (SCHEDULE): %s\n", err.Error())
			return
		}
		s.schedule[hour] = true
	}
	if len(s.schedule) > 24 {
		err = errors.New("too many hours in schedule")
		fmt.Printf("failed to parse array of integers from environment variable (SCHEDULE): %s\n", err.Error())
		return
	}

	tz := os.Getenv("TZ")
	if tz != "" {
		s.tz = defaultTimezone
	}

	return
}

// controlBasedOnThreshold controls heating based on threshold
func (s state) controlBasedOnThreshold() (err error) {
	now := time.Now()
	price, err := s.sp.GetPrice(now)
	if err != nil {
		fmt.Printf("failed to control based on threshold: %s", err.Error())
		return err
	}

	fmt.Printf("control based on threshold (%.2f)\n", s.threshold)
	fmt.Printf("hourly price [%s]: %.2f\n", time.Now().Format(time.RFC822), price)

	if price <= s.threshold {
		// heating ON / NORMAL mode (price is lower than the threshold)
		fmt.Printf("Heating ON: price lower than the threshold: %0.2f (threshold: %0.2f)\n", price, s.threshold)

		err = s.cs.SwitchOff()
		if err != nil {
			fmt.Printf("failed to turn heat pump on: %s\n", err.Error())
			return err
		}
	} else {
		// heating OFF / ROOM LOWERING mode
		fmt.Printf("Heating OFF: price higher than the threshold: %0.2f (threshold: %0.2f)\n", price, s.threshold)
		err = s.cs.SwitchOn()
		if err != nil {
			fmt.Printf("failed to turn heat pump off / room lowering mode: %s\n", err.Error())
		}
	}
	return nil
}

// controlBasedOnActiveHours controls heating based on activeHours (and maxPrice if set)
func (s state) controlBasedOnActiveHours() (err error) {
	now := time.Now()
	price, err := s.sp.GetPrice(now)
	if err != nil {
		fmt.Printf("failed to control based on activeHours: %s", err.Error())
		return err
	}

	if spotprice.IsCheapestHour(now.Hour(), s.sp.CheapestHours(s.activeHours)) {
		// This is one of the cheapest hours
		if price > s.maxPrice {
			fmt.Printf("Heating OFF: this is one of the %d cheapest hours: %0.2f, but price is higher than maxPrice (%0.2f)\n", s.activeHours, price, s.maxPrice)
		} else {
			// Heating ON / NORMAL mode (this is one of the cheapest hours)
			fmt.Printf("Heating ON: this is one of the %d cheapest hours: %0.2f\n", s.activeHours, price)
			err = s.cs.SwitchOff()
			if err != nil {
				fmt.Printf("failed to turn heat pump on: %s\n", err.Error())
				return err
			}
		}
	} else {
		fmt.Printf("Heating OFF: this is not one of the %d cheapest hours\n", s.activeHours)
		err = s.cs.SwitchOn()
		if err != nil {
			fmt.Printf("failed to turn heat pump off / room lowering mode: %s\n", err.Error())
		}
	}

	return nil
}

// controlBasedOnThresholdAndActiveHours controls heating based on threshold and activeHours (and maxPrice if set)
func (s state) controlBasedOnThresholdAndActiveHours() (err error) {
	now := time.Now()
	price, err := s.sp.GetPrice(now)
	if err != nil {
		fmt.Printf("failed to control based on threshold and activehours: %s", err.Error())
		return err
	}

	fmt.Printf("control based on threshold (%.2f) and active hours (%d)\n", s.threshold, s.activeHours)
	fmt.Printf("hourly price [%s]: %.2f\n", time.Now().Format(time.RFC822), price)

	if price <= s.threshold {
		// heating ON / NORMAL mode (price is lower than the threshold)
		err = s.cs.SwitchOff()
		if err != nil {
			fmt.Printf("failed to turn heat pump on: %s\n", err.Error())
			return err
		}
	} else {
		// price is higher than the threshold
		if s.activeHours > 0 {
			if spotprice.IsCheapestHour(now.Hour(), s.sp.CheapestHours(s.activeHours)) {
				// This is one of the cheapest hours
				if price > s.maxPrice {
					fmt.Printf("Heating OFF: this is one of the %d cheapest hours: %0.2f, but price is higher than maxPrice (%0.2f)\n", s.activeHours, price, s.maxPrice)
				} else {
					// Heating ON / NORMAL mode (this is one of the cheapest hours)
					fmt.Printf("Heating ON: price higher than threshold but this is one of the %d cheapest hours: %0.2f\n", s.activeHours, price)
					err = s.cs.SwitchOff()
					if err != nil {
						fmt.Printf("failed to turn heat pump on: %s\n", err.Error())
						return err
					}
				}
			} else {
				fmt.Printf("Heating OFF: price higher than threshold and this is not one of the %d cheapest hours\n", s.activeHours)
				// heating OFF / ROOM LOWERING mode
				err = s.cs.SwitchOn()
				if err != nil {
					fmt.Printf("failed to turn heat pump off / room lowering mode: %s\n", err.Error())
				}
			}
		} else {
			// heating OFF / ROOM LOWERING mode
			err = s.cs.SwitchOn()
			if err != nil {
				fmt.Printf("failed to turn heat pump off / room lowering mode: %s\n", err.Error())
			}
		}
	}
	return nil
}

// controlBasedOnCron controls heating based on cron
func (s state) controlBasedOnSchedule() (err error) {
	now := time.Now()
	price, _ := s.sp.GetPrice(now)

	fmt.Printf("control based on schedule\n")

	if s.schedule[now.Hour()] == true {
		// Heating ON / NORMAL mode
		fmt.Printf("Heating ON: schedule (price: %0.2f)\n", price)
		err = s.cs.SwitchOff()
		if err != nil {
			fmt.Printf("failed to turn heat pump on: %s\n", err.Error())
			return err
		}
	} else {
		// heating OFF / ROOM LOWERING mode
		fmt.Printf("Heating OFF: schedule (price: %0.2f)\n", price)
		err = s.cs.SwitchOn()
		if err != nil {
			fmt.Printf("failed to turn heat pump off / room lowering mode: %s\n", err.Error())
		}

	}
	return nil
}
